smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/lang/DocumentationPreprocessor.kt

```diff
@@ -21,17 +26,226 @@ class DocumentationPreprocessor : KotlinIntegration {
+                "a" -> {
+                    writeBufferedAnchor()
+                }
+                "ul", "ol" -> {
+                    sublistDedent()
+                    text += if (node.parent()?.nodeName() == "body") "\n\n" else ""
+                }
+                "code", "pre" -> {
+                    text += PREFORMAT_MARKER
+                }
+                "b", "strong" -> {
+                    text += BOLD_MARKER
+                }
+                "i", "em" -> {
+                    text += ITALIC_MARKER
+                }
+                "li", "fullname", "dt" -> {
+                    text += "\n"
+                }
+            }
+        }
+
+        private fun writeBufferedAnchor() {
+            // Model docs will sometimes contain an anchor without the href. At that point there's no real way of
+            // knowing to what it refers, nor can we guarantee a valid link just by bracketing it.
+            text += if (bufferedAnchorHref != "") {
+                "[$bufferedAnchorText]($bufferedAnchorHref)"
+            } else {
+                bufferedAnchorText
+            }
+
+            bufferedAnchorHref = ""
+            bufferedAnchorText = ""
+        }
+
+        private fun sublistIndent() {
+            listPrefix += SUBLIST_INDENT
+        }
+
+        private fun sublistDedent() {
+            listPrefix = listPrefix.dropLast(SUBLIST_INDENT.length)
+        }
+    }
+}
+
+/**
+ * Jsoup will preserve newlines between elements as blank text nodes. These have zero bearing on the content of the
+ * document to begin with and only serve to complicate traversal.
+ */
+private fun Node.stripBlankTextNodes() {
+    if (this is TextNode && this.isBlank) {
+        this.remove()
+        return
+    }
+
+    val childNodes = this.childNodes()
+    if (childNodes.isNotEmpty()) {
+        childNodes.forEach(Node::stripBlankTextNodes)
+    }
+}
+
+private fun Node.hasAncestor(predicate: (Node) -> Boolean): Boolean {
+    if (!this.hasParent()) return false
+
+    val parent = this.parent() as Node
+    return predicate(parent) || parent.hasAncestor(predicate)
+}
+
+private fun Node.isList() =
+    this.nodeName() == "ul" || this.nodeName() == "ol"
+
+private fun TextNode.markdownText() =
+    this.text()
+        // Replace square brackets with escaped equivalents so that they are not rendered as invalid Markdown
+        // links.
+        .replace("[", "&#91;")
+        .replace("]", "&#93;")
+
+/**
+ * Operates on all substrings that fall within the provided section delimiters. Returns a new string where all
+ * substrings enclosed as specified have been modified according to the provided transform.
+ */
+private fun String.applyWithin(start: String, end: String, transform: (String) -> String): String {
+    val substringStart = this.indexOf(start) + start.length
+    if (substringStart == -1) return this
+
+    val substringEnd = this.indexOf(end, substringStart)
+    if (substringEnd == -1) return this
```

Comment:
**Question**: Is this intended to operate on nested constructs? If so, I think it might be broken because it assumes that `end` will be found before any subsequent `start` instances.

Reply:
It definitely isn't built to handle that. That said, within the specific context of what it's accomplishing in this module (ie. sanitizing HTML within preformat) any documentation that came to us like that would be very broken on its own(in a different way than just having preformat blocks with unescaped `<>&` characters).

eg. Consider the block
```
<code>
<shouldEscapeTheseBrackets>
someFunction();<code></code>
<shouldEscapeTheseBracketsToo>
</code>
```

If those inner tags aren't escaped, the structure of the document is basically destroyed past the first closing tag. The existing implementation would read the block to operate on as 
```
<code>
<shouldEscapeTheseBrackets>
someFunction();<code></code>
```
and the HTML we feed into the parser afterward would be garbage.

I think there's a balance to be struck here, given that we're making a concerted effort to clean up what is essentially malformed HTML to begin with. This may all be moot if there ends up being a regex solution.

Assuming no regex, for this particular application, we'd really want to assume that the inner tags SHOULD be rendered with the tags escaped. Therefore we'd ignore any nested delimiters and let that text be mutated as part of the enclosing pair. **So I think the function should definitely be modified to not get tripped up on nested structures, but it should basically ignore them (and even then, you're still gonna get to operate on that text within the nested structure in your transform, you're just not gonna get a standalone callback for it).**

There is also the copout route of just throwing up hands if we encounter a nesting like this, but if we can detect it we might as well handle it sensibly. Curious what you guys think.